### PR TITLE
RDK-35522: Add Unit Tests for DeviceDiagnostics using Google Test Fra…

### PR DESCRIPTION
**Pull request by mateusz-adamski-red**
…mework

Signed-off-by: mateusz.adamski <mateusz.adamski@consult.red>

### CODE REVIEW

**Comment by npoltorapavlo** [comment]
please build w/ `ENABLE_ERM`, call Initialize / Deinitialize, add a test for the event (mock onDecoderStatusChange similar to https://github.com/rdkcentral/rdkservices/blob/sprint/2203/RdkServicesTest/Mocks/PersistentStoreMock.h#L33). Document options in line:
```
./Scripts/build.sh [-j<number>] [-D]

options:
  -D          debug mode
  ...
```

File: RdkServicesTest/Tests/DeviceDiagnosticsTest.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+**/
+
+#include "gtest/gtest.h"
+
+#include <thread>
+#include <chrono>
```
Comment (npoltorapavlo) [inline RdkServicesTest/Tests/DeviceDiagnosticsTest.cpp:23]: please remove

File: RdkServicesTest/Tests/DeviceDiagnosticsTest.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+TEST_F(DeviceDiagnosticsTest, ShouldGetEvent) {
+    EXPECT_EQ(string(""), deviceDiagnostic_->Initialize(nullptr));
+
+    EXPECT_CALL(*deviceDiagnostic_, onDecoderStatusChange(1));
+    // Wait for incoming event
+    std::this_thread::sleep_for(std::chrono::milliseconds(2));
+
+    deviceDiagnostic_->Deinitialize(nullptr);
+}
```
Comment (npoltorapavlo) [inline RdkServicesTest/Tests/DeviceDiagnosticsTest.cpp:66]: please remove

File: RdkServicesTest/Source/essos-resmgr.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#pragma once
```
Comment (npoltorapavlo) [inline RdkServicesTest/Source/essos-resmgr.h:1]: please remove this file

File: RdkServicesTest/Scripts/run.sh [hunk 4]
```diff
@@ -27,9 +33,12 @@ valgrind \
 --leak-check=yes \
 --show-reachable=yes \
 --track-fds=yes \
---fair-sched=try \
+--fair-sched=try \
```
Comment (npoltorapavlo) [inline RdkServicesTest/Scripts/run.sh:36]: please revert

File: RdkServicesTest/Scripts/build.sh [hunk 5]
```diff
@@ -144,8 +197,12 @@ buildAndInstallThunderInterfaces
 
 checkWPEFramework
 
+changeDeviceDiagnostics
+
 buildAndInstallRdkservices
 
+cleanupDeviceDiagnostics
+
```
Comment (npoltorapavlo) [inline RdkServicesTest/Scripts/build.sh:205]: please remove

File: RdkServicesTest/Scripts/build.sh [hunk 6]
```diff
@@ -144,8 +197,12 @@ buildAndInstallThunderInterfaces
 
 checkWPEFramework
 
+changeDeviceDiagnostics
+
```
Comment (npoltorapavlo) [inline RdkServicesTest/Scripts/build.sh:201]: please remove

File: RdkServicesTest/Scripts/build.sh [hunk 7]
```diff
@@ -99,7 +104,22 @@ buildAndInstallThunderInterfaces() {
...
+  cd "${ROOT}";
+  sed -i 's/#pragma once/#pragma once\n#define ENABLE_ERM/' ../DeviceDiagnostics/DeviceDiagnostics.h
+  # to speed up a test change the limit of sleeping before first event (normally need to wait 30 sec)
+  sed -i 's/.tv_sec = 30, .tv_nsec = 0/.tv_sec = 0, .tv_nsec = 1000/' ../DeviceDiagnostics/DeviceDiagnostics.cpp
+}
+
+cleanupDeviceDiagnostics() {
+  cd "${ROOT}";
+  sed -i '/^#pragma once$/{$!{N;s/^#pragma once\n#define ENABLE_ERM$/#pragma once/;ty;P;D;:y}}' ../DeviceDiagnostics/DeviceDiagnostics.h
+  sed -i 's/.tv_sec = 0, .tv_nsec = 1000/.tv_sec = 30, .tv_nsec = 0/' ../DeviceDiagnostics/DeviceDiagnostics.cpp
```
Comment (npoltorapavlo) [inline RdkServicesTest/Scripts/build.sh:122]: please remove

File: DeviceDiagnostics/DeviceDiagnostics.cpp [hunk 8]
```diff
@@ -71,8 +71,6 @@ namespace WPEFramework
         /* virtual */ const string DeviceDiagnostics::Initialize(PluginHost::IShell* service)
         {
 #ifdef ENABLE_ERM
-            int ret;
-
```
Comment (npoltorapavlo) [inline DeviceDiagnostics/DeviceDiagnostics.cpp:75]: please revert

File: DeviceDiagnostics/DeviceDiagnostics.h [hunk 9]
```diff
@@ -42,7 +42,7 @@ namespace WPEFramework {
             int getConfiguration(const std::string& postData, JsonObject& response);
             uint32_t getAVDecoderStatus(const JsonObject& parameters, JsonObject& response);
             int getMostActiveDecoderStatus();
-            void onDecoderStatusChange(int status);
+            virtual void onDecoderStatusChange(int status);
```
Comment (npoltorapavlo) [inline DeviceDiagnostics/DeviceDiagnostics.h:45]: please revert

File: RdkServicesTest/README.md [hunk 10]
```diff
@@ -11,10 +11,18 @@ Read the [GoogleTest User’s Guide](https://google.github.io/googletest/).
...
+
+## Build options
 
-Make sure you have packages _python3 python3-pip libsqlite3-dev_.
+```
+$./Scripts/build.sh [-j<number>] [-D]
+
+where:
+    -j  number of threads (default maximum available threads)
+    -D  debug mode (default Release)"
```
Comment (npoltorapavlo) [inline RdkServicesTest/README.md:24]: can you add this info to "How to build"?

File: RdkServicesTest/README.md [hunk 11]
```diff
@@ -11,10 +11,18 @@ Read the [GoogleTest User’s Guide](https://google.github.io/googletest/).
 cd RdkServicesTest
 ./Scripts/build.sh
 ```
-
 This builds Thunder, ThunderInterfaces, Rdkservices and installs them in thunder/install.
+Make sure you have packages python3 python3-pip libsqlite3-dev, curl libcurl4-openssl-dev
```
Comment (npoltorapavlo) [inline RdkServicesTest/README.md:15]: please just add "libcurl4-openssl-dev". without other changes

File: RdkServicesTest/Mocks/DeviceDiagnosticsMock.h [hunk 12]
```diff
@@ -0,0 +1,33 @@
+/*
```
Comment (npoltorapavlo) [inline RdkServicesTest/Mocks/DeviceDiagnosticsMock.h:1]: please remove this file

File: RdkServicesTest/Tests/DeviceDiagnosticsTest.cpp [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+**/
+
+#include "gtest/gtest.h"
+
+#include <thread>
+#include <chrono>
+
+#include "DeviceDiagnosticsMock.h"
```
Comment (npoltorapavlo) [inline RdkServicesTest/Tests/DeviceDiagnosticsTest.cpp:25]: please remove

File: RdkServicesTest/Tests/DeviceDiagnosticsTest.cpp [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+class DeviceDiagnosticsTest : public ::testing::Test {
+public:
+    DeviceDiagnosticsTest():
+        deviceDiagnostic_(Core::ProxyType<DeviceDiagnosticsMock>::Create()),
+        handler_(*deviceDiagnostic_),
+        connection_(1, 0) {
+    }
+
+protected:
+    Core::ProxyType<DeviceDiagnosticsMock> deviceDiagnostic_;
```
Comment (npoltorapavlo) [inline RdkServicesTest/Tests/DeviceDiagnosticsTest.cpp:38]: please use DeviceDiagnostics

File: DeviceDiagnostics/DeviceDiagnostics.cpp (hunk 8 above)
Comment (mateusz-adamski-red) [reply DeviceDiagnostics/DeviceDiagnostics.cpp:75]: This variable is never used, I just make cleanup in code

**Comment by npoltorapavlo** [comment]
looks good

**Review by anand-ky — APPROVED** [review]

**Pull request merged**